Week 1 & Week Projects - Gen AI BootCamp (Joshua Oluoch)

TheTopDeveloper opened the pull request:
### Hidden Gems World Travel Guide

A lightweight Retrieval-Augmented Generation (RAG) chat that answers travel questions about “hidden gem” sites using locally generated Markdown guides. It retrieves passages from a small knowledge base (15 countries; up to 10 sites per country), composes answers with OpenAI gpt-5-nano, and quietly evaluates/retries with Anthropic claude-sonnet-4-5 until quality is acceptable.

#### How it works
- Retrieval: OpenAI embeddings over `hidden_gems_output/*_guide.md` with cosine search.
- Answering: OpenAI `gpt-5-nano`, grounded strictly on retrieved context.
- Evaluation: Anthropic `claude-sonnet-4-5` checks relevance/grounding/robustness; auto-retries are hidden from the user.
- UX: Gradio chat with info header listing supported countries and per‑site fields you can ask about.

#### Key features
- RAG over your own Markdown content (no SDKs; pure `requests`).
- Clear dataset scope: 15 countries, ≤10 sites per country.
- Country-aware behavior: flags non-covered countries; avoids false “not covered” when present.
- Clean information header: canonicalized, de-duplicated per-site fields (e.g., Best time to visit, Transportation access, Safety information).

#### Screenshots
- Screenshot 1: Landing header (countries covered + per-site fields).
- 
<img width="757" height="356" alt="Screenshot1" src="https://github.com/user-attachments/assets/f7c41d59-79db-4253-a73d-72cb46620634" />

- Screenshot 2: Sample chat showing retrieval-backed Q&A and follow-ups.

<img width="747" height="229" alt="Screenshot2" src="https://github.com/user-attachments/assets/143723d7-da14-4f69-a3b7-39c7fca2ae74" />

#### Deployment
The app is live on HuggingFace Space on https://huggingface.co/spaces/joshuaoluoch88/hidden_gems_world_travel_guide

<img width="875" height="460" alt="image" src="https://github.com/user-attachments/assets/5ac2340b-c93d-4043-884b-83439625a8a8" />


#### Achievements
- Built an end-to-end RAG with strict grounding and silent auto-retries.
- Robust country detection + normalized UI fields for clarity.
- Zero SDK dependency; OpenAI/Anthropic via REST calls.

#### Lessons learned
- Main lesson: Agentic AI — using one LLM to evaluate and improve
- Small RAGs need guardrails (coverage checks + evaluator) to reduce hallucinations.
- Normalizing metadata (aliases/truncations) improves usability and searchability.
- Including recent chat turns in retrieval keeps follow-ups on-topic.

#### Limitations
- Knowledge base intentionally small (15 countries; up to 10 sites each).
- Dependent on external APIs and rate limits for answering/evaluation.

#### Roadmap
- Add snippet-level citations and site IDs.
- Expand KB + country/site filters.
- Cache embeddings/retrieval results.
- Optional streaming responses.

#### Tech stack
- Python, Gradio, NumPy, requests, python-dotenv
- Models: OpenAI `gpt-5-nano`; Anthropic `claude-sonnet-4-5`
- Env vars: `OPENAI_API_KEY` (required), `ANTHROPIC_API_KEY` (optional), `HF_TOKEN` for Spaces deployment

ed-donner commented:
Hey Joshua,
This is really great stuff - thank you - I will mark you complete for week 2..
Hey before I merge this.. it's 5,000+ lines of code, with (for example) lots of detail in the travel guides. Would it be possible to do some housekeeping - line maybe only include skeleton versions of these files? Thanks so much

TheTopDeveloper commented:
I have removed the hidden_gems_output folder and added instructions in the README.md file on generating the knowledge base. This should reduce the codebase volume substantially.

ed-donner commented:
That's great - many thanks - a powerful project
